File: freight-triage-polyglot-ledger-recovery/environment/native/src/hash/elf_hash.cpp
#include "freight/hashes.h"

namespace freight {

// elf_hash over raw bytes.
uint64_t hash_elf_hash(const std::string& data) {
  uint32_t state = 0u;
  for (size_t i = 0; i < data.size(); ++i) {
    state = (state << 4) + static_cast<uint8_t>(data[i]);
    uint32_t high = state & 0xF0000000u;
    if (high != 0u) {
      state ^= high >> 24;
    }
    state &= ~high;
  }
  return static_cast<uint64_t>(state);
}

}  // namespace freight

File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/hash/ElfHash.java
package com.freight.hash;

/** elf_hash over raw bytes. */
public final class ElfHash implements HashAlgorithm {

    @Override
    public String name() {
        return "elf_hash";
    }

    @Override
    public long apply(byte[] data) {
        long state = 0L;
        for (int i = 0; i < data.length; i++) {
            state = ((state << 4) + (data[i] & 0xFF)) & 0xFFFFFFFFL;
            long high = state & 0xF0000000L;
            if (high != 0L) {
                state ^= high >>> 24;
            }
            state &= (~high) & 0xFFFFFFFFL;
        }
        return state;
    }
}
